File: api/src/vibecanvas_api/browser/envelope.py
"""Shared envelope for browser ping, command, and observation messages."""
from __future__ import annotations
import json

_REQUIRED = ("kind", "id", "channel", "transport")

def encode(kind: str, *, id: str, channel: str, transport: str,
           data: dict | None = None, producer: str | None = None) -> str:
    return json.dumps({"v": 1, "kind": kind, "id": id, "channel": channel,
                       "transport": transport, "data": data, "producer": producer},
                      separators=(",", ":"))

def decode(raw: str) -> dict:
    try:
        d = json.loads(raw)
    except Exception as e:
        raise ValueError(f"malformed envelope: {e}") from e
    if not isinstance(d, dict) or any(k not in d for k in _REQUIRED):
        raise ValueError("envelope missing required fields")
    return d

File: extension/src/shared/envelope.ts
/**
 * The single shared message envelope (§5.4) — TypeScript twin of
 * `api/src/vibecanvas_api/browser/envelope.py`. Encode/decode MUST produce and
 * accept the same JSON shape as the Python side so the two ends interoperate:
 *
 *   { v: 1, kind, id, channel, transport, data, producer }
 *
 * `data` and `producer` default to `null` (the wire form of Python's `None`).
 * Ping, command, and observation traffic share this envelope and WebSocket
 * client.
 */

export type Envelope = {
  v: 1;
  kind: string;
  id: string;
  channel: string;
  transport: string;
  data: unknown;
  producer: string | null;
};

/** Fields that MUST be present for a frame to be a valid envelope. */
const REQUIRED = ["kind", "id", "channel", "transport"] as const;

export function encode(
  kind: string,
  o: {
    id: string;
    channel: string;
    transport: string;
    data?: unknown;
    producer?: string;
  },
): string {
  return JSON.stringify({
    v: 1,
    kind,
    id: o.id,
    channel: o.channel,
    transport: o.transport,
    data: o.data ?? null,
    producer: o.producer ?? null,
  });
}

export function decode(raw: string): Envelope {
  let d: unknown;
  try {
    d = JSON.parse(raw);
  } catch {
    throw new Error("malformed envelope: not JSON");
  }
  if (
    typeof d !== "object" ||
    d === null ||
    REQUIRED.some((k) => !(k in (d as Record<string, unknown>)))
  ) {
    throw new Error("envelope missing required fields");
  }
  return d as Envelope;
}
